### PR TITLE
Refactor Cache Service

### PR DESCRIPTION
**Pull request by AugustoL**
# Description

- Add processProposal that gets the immutable + mutable data if the proposal is new, and if not it gets only the mutable data and updates the proposal in the cache.
- Add processScheme fucntion that gets the immutable + mutable data if the scheme is new, and if not it gets only the mutable data and updates the scheme in the cache. Also if the scheme is new it gets the proposal creation events and adds the new proposals.
- Refactor code form cacheService to utils, adding more utils function and removing duplicated code.
- Removed some unnecesary varibales requested as parameters.

I tested the xdai cache generated againts the latest we build for xdai and there are very few differences, in fact, the cache built from this PR is more up to date.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

# Checklist:

- [x] My code follows the existing style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any UI changes have been tested and made responsive for mobile views

### CODE REVIEW

File: src/services/CacheService.ts [hunk 1]
```diff
@@ -711,35 +668,65 @@ export default class UtilsService {
...
-            start: callsResponse.decodedReturnData[4],
-            isOwner: callsResponse.decodedReturnData[5],
-            revocable: callsResponse.decodedReturnData[6],
+            beneficiary: callsResponse.decodedReturnData[0][0],
+            cliff: callsResponse.decodedReturnData[1][0],
+            duration: callsResponse.decodedReturnData[2][0],
+            owner: callsResponse.decodedReturnData[3][0],
+            start: callsResponse.decodedReturnData[4][0],
+            isOwner: callsResponse.decodedReturnData[5][0],
+            revocable: callsResponse.decodedReturnData[6][0],
```
Comment (MiltonTulli) [inline src/services/CacheService.ts:730]: I think we are not using these 2 keys (`isOwner` `revocable`) in dxvote and we can avoid storing them in the cache unless we know that we will use them in the closer future.

File: src/services/CacheService.ts [hunk 2]
```diff
@@ -758,1318 +745,116 @@ export default class UtilsService {
...
+            // If not, it means that the proposal was created in the current block range when processing the scheme,
+            // So there is no need to process it again.
+            networkCache.proposals[proposalId].creationEvent.blockNumber <
+              fromBlock &&
+            // This condition check that the proposal is active
+            (networkCache.proposals[proposalId].stateInVotingMachine >
+              VotingMachineProposalState.Executed ||
+              networkCache.proposals[proposalId].stateInScheme ===
+                WalletSchemeProposalState.Submitted)
+          );
```
Comment (MiltonTulli) [inline src/services/CacheService.ts]: Just thinking of making the cacheService more readable and debuggable, I suggest:
```suggestion
    const proposal = networkCache.proposals[proposalId];
    const scheme = networkCache.schemes[proposal?.scheme];
    return (
      // Only update proposals for registered schemes
      scheme?.registered &&
      // This condition check that the proposal already existed in the current block range.
      // If not, it means that the proposal was created in the current block range when processing the scheme,
      // So there is no need to process it again.
      proposal?.creationEvent.blockNumber < fromBlock &&
      // This condition check that the proposal is active
      (proposal?.stateInVotingMachine > VotingMachineProposalState.Executed ||
        proposal?.stateInScheme === WalletSchemeProposalState.Submitted)
    );
```

File: src/services/CacheService.ts [hunk 3]
```diff
@@ -2150,4 +935,957 @@ export default class UtilsService {
...
+  // If the proposal does not exist in the cache it gets the immutable + mutable data
+  // The proposal already exists it gets the mutable data only
+  async processProposal(
+    proposalId: string,
+    networkCache: DaoNetworkCache,
+    networkContracts: NetworkContracts,
+    web3: Web3,
+    fromBlock: number,
+    toBlock: number,
+    creationEvent?: any
```
Comment (MiltonTulli) [inline src/services/CacheService.ts]: Should we add types for this? (BlockchainEvent?)
```
{
event,
signature,
address,
transactionHash,
blockNumber,
timestamp,
transactionIndex,
logIndex,
}

File: src/utils/helpers.ts [hunk 4]
```diff
@@ -87,3 +87,20 @@ export const appendEthAPIKey = (
 export function escapeRegExp(string: string): string {
   return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
 }
+
+export async function retryPromise(
```
Comment (MiltonTulli) [inline src/utils/helpers.ts:91]: Should we add a `maxTries` arg?

**Review by MiltonTulli — DISMISSED** [review]
Didn't test it but the code looks good to me!. Just some comments.

File: src/services/CacheService.ts (hunk 2 above)
Comment (AugustoL) [reply src/services/CacheService.ts]: https://github.com/DXgovernance/dxvote/pull/795/commits/ecf96b5ea0948d9d1ea208dc8148941cf013f97b

File: src/utils/helpers.ts (hunk 4 above)
Comment (AugustoL) [reply src/utils/helpers.ts:91]: https://github.com/DXgovernance/dxvote/pull/795/commits/97e931ccfee1013cce4e716dd240c662c533eb18

File: src/services/CacheService.ts (hunk 3 above)
Comment (AugustoL) [reply src/services/CacheService.ts]: https://github.com/DXgovernance/dxvote/pull/795/commits/06704c688de6d7413117543c2bffce88574bada3

File: src/services/CacheService.ts (hunk 1 above)
Comment (AugustoL) [reply src/services/CacheService.ts:730]: No need to change this now, this is one of the first things that can be moved to hooks from the cache context, also we might want that fields in a future.

**Comment by rossneilson** [comment]
Testing this I am seeing some pretty large differences from develop, both loaded at the same time from remote cache
1 boosted vs 4 boosted
0 pre boost vs 5 pre boost
2 queued vs 4 queued
Looks like the missing ones don't exist
Preview:
Screenshot 2022-04-21 at 12 53 26
Develop:
Screenshot 2022-04-21 at 12 53 44

**Comment by AugustoL** [comment]
> Testing this I am seeing some pretty large differences from develop, both loaded at the same time from remote cache 1 boosted vs 4 boosted 0 pre boost vs 5 pre boost 2 queued vs 4 queued Looks like the missing ones don't exist Preview: Screenshot 2022-04-21 at 12 53 26 Develop: Screenshot 2022-04-21 at 12 53 44

hmmm, going to debug, thx

**Comment by AugustoL** [comment]
@rossneilson should be fixed now. I comapred a resetted cache and only mutable values changed in some proposals, mainly boosts and bounty redeems.

**Review by rossneilson — APPROVED** [review]

**Pull request merged**